datasource/src/main/scala/quasar/physical/s3/impl/isResource.scala

```diff
@@ -50,14 +51,15 @@ object isResource {
       false.pure[F]
     } else {
       // Don't use the metadata, just check the request status
-      client.status(request) >>= {
-        case Status.Ok => true.pure[F]
-        case Status.PartialContent => true.pure[F]
-        case Status.NotFound => false.pure[F]
-        case Status.RangeNotSatisfiable => false.pure[F]
-        case Status.Forbidden => Sync[F].raiseError(new Exception(s"Permission denied. Make sure you have access to the configured bucket"))
-        case s => Sync[F].raiseError(new Exception(s"Unexpected status returned during `isResource` call: $s"))
-      }
+      sign(request) >>= (r =>
+        client.status(r) >>= {
+          case Status.Ok => true.pure[F]
+          case Status.PartialContent => true.pure[F]
```

Comment:
Should we be alerting and/or failing in the case that content was partial? Otherwise wouldn't be silently continue on?

Reply:
S3 technically does not support directories, only prefixes. However, it is common for tools that use S3 to create "directories" by creating empty object with a prefix-only name. Even Amazon's own S3 UI does this.

So the idea here is to identify these "directories" (empty objects) by attempting to download the [first byte of it](https://github.com/slamdata/quasar-s3/pull/56/files/7da152a7bd70f9f24bc9ce277043a92008a1628a#diff-3f9bf430ba160bbebb4dee3702507637R48). If we get a `404`, that means the object does not exist. If get `416 Range Not Satisfiable`, that means this an empty object (that is, a "directory"). For these two cases we return `false`.

`206 Partial Content` or `200 OK` means this object does exist and has contents in it, so we return true.

Any other status is abnormal and hence we `raiseError`.